### PR TITLE
Aws implement

### PR DESCRIPTION
**Pull request by dtsehl**
### New Feature Submissions:
* [X] Have you added an explanation of what your changes do?
* [X] Have you linted your code locally prior to submission?

### Testing:
* [X] Have you written new tests for your core changes, as applicable?
* [X] Have you successfully ran tests with your changes locally?
* [X] All new and existing tests passed?

**Test coverage** = 100%
--------------------------------------------------------------------------------
### Change Log:
- Utilize the aws-sdk-s3 gem
- Implement S3able module to handle all interaction with S3
- Made a migration to add key column to ApiImage table which is a string
- We are now just passing one ApiImage object
- Add more environment variables to application.yml file - AWS bucket

### CODE REVIEW

File: app/controllers/images_controller.rb [hunk 1]
```diff
@@ -1,12 +1,16 @@
...
   include Glitchable
+  include S3able
+
   def create
-    image = current_user.api_images.create!(image_params)
-    download = Down.download(image.url)
-    image.original.attach(io: download, filename: "#{image.url}.jpg")
-    basic_glitch(image)
-    redirect_to search_show_path(image)
+    api_image = current_user.api_images.create!(image_params)
```
Comment (dtsehl) [inline app/controllers/images_controller.rb:7]: This controller action is still quite fat, may need some refactoring down the line.

File: app/controllers/images_controller.rb [hunk 2]
```diff
@@ -1,12 +1,16 @@
...
+    download = Down.download(api_image.url)
+    api_image.original.attach(io: download, filename: "#{api_image.url}.jpg")
+    basic_glitch(api_image)
+    api_image.attach_key(upload_image)
```
Comment (dtsehl) [inline app/controllers/images_controller.rb:11]: The return value of upload_image is a unique key which can then be called to get the location of the image stored by S3.

File: app/models/api_image.rb [hunk 3]
```diff
@@ -6,4 +6,8 @@ class ApiImage < ApplicationRecord
 
   has_one_attached :original
   has_one_attached :glitched
+
+  def attach_key(key)
```
Comment (janegreene) [inline app/models/api_image.rb:10]: S3 needs the files to land with a unique name so we have used SecureRandom uuid to generate this and attach it to the object.

File: app/modules/s3able.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  def open_connection
+    credentials = Aws::Credentials.new(ENV['AWS_ACCESS_KEY_ID'], ENV['AWS_SECRET_ACCESS_KEY'])
+    Aws::S3::Client.new(region: 'us-east-1', credentials: credentials)
+  end
+
+  def upload_image
+    connection = open_connection
+    key = generate_key
+    File.open(ENV['PICTURE_PATH'], 'rb') do |file|
+      connection.put_object(bucket: ENV['AWS_BUCKET'], key: key, body: file)
```
Comment (kathytbui) [inline app/modules/s3able.rb:16]: This was a lot of reading - thanks for figuring it out!

File: app/views/search/show.html.erb [hunk 5]
```diff
@@ -15,7 +14,7 @@
     <div class="card embed-responsive-4by3" style="width: 40rem; margin-top: 10px; background: rgba(0, 0, 0, 0.8);">
       <div class="card-body" style="color:#b3b3b3; margin: 10px;">
         <section class="image">
-          <%= image_tag @image.glitched, :class => "glitched-image embed-responsive-item img-fluid" %>
+          <%= image_tag "https://#{ENV['AWS_BUCKET']}.s3.amazonaws.com/#{@api_image.key}", :class => "glitched-image embed-responsive-item img-fluid" %>
```
Comment (kathytbui) [inline app/views/search/show.html.erb:17]: Thanks for figuring this one out, Jane.

**Review by sciencefixion — APPROVED** [review]
This is masterful. A monument to your amazing effort.

**Pull request merged**